client/ping.go

```diff
@@ -13,18 +13,15 @@ func (cli *Client) Ping(ctx context.Context) (types.Ping, error) {
 		return ping, err
 	}
 	serverResp, err := cli.doRequest(ctx, req)
-	if err != nil {
-		return ping, err
-	}
-	defer ensureReaderClosed(serverResp)
+	if serverResp.header != nil {
```

Comment:
`cli.doRequest` does two things - first it issues the actual HTTP request with `ctxhttp.Do`, then it checks the status code to determine whether the request was successful.

Could we separate those two things, so we can distinguish a HTTP-level problem from an application-level problem, and only check the headers when the HTTP request went through? That would avoid using `serverResp` even when an error is returned, which can be a problematic pattern.